radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/TokensConstraintScryptV1.java

```diff
@@ -53,18 +58,89 @@ public void main(Loader os) {
 	}
 
 	private void registerParticles(Loader os) {
-		os.particle(
-			TokenResource.class,
-			ParticleDefinition.<TokenResource>builder()
-				.staticValidation(TokenDefinitionUtils::staticCheck)
-				.build()
+		os.substate(
+			new SubstateDefinition<>(
+				TokenResource.class,
+				Set.of(SubstateTypeId.TOKEN_DEF.id()),
+				(b, buf) -> {
+					var rri = REFieldSerialization.deserializeREAddr(buf);
+					var type = buf.get();
+					final UInt256 supply;
+					final ECPublicKey minter;
+					if (type == 0) {
+						supply = null;
+						minter = null;
+					} else if (type == 1) {
+						supply = null;
+						minter = REFieldSerialization.deserializeKey(buf);
+					} else if (type == 2) {
+						supply = REFieldSerialization.deserializeUInt256(buf);
+						minter = null;
+					} else {
+						throw new DeserializeException("Unknown token def type " + type);
+					}
+					var name = REFieldSerialization.deserializeString(buf);
+					var description = REFieldSerialization.deserializeString(buf);
+					var url = REFieldSerialization.deserializeUrl(buf);
+					var iconUrl = REFieldSerialization.deserializeUrl(buf);
+					return new TokenResource(rri, name, description, iconUrl, url, supply, minter);
+				},
+				(s, buf) -> {
+					buf.put(SubstateTypeId.TOKEN_DEF.id());
+					REFieldSerialization.serializeREAddr(buf, s.getAddr());
+					s.getSupply().ifPresentOrElse(
+						i -> {
+							buf.put((byte) 2);
+							buf.put(i.toByteArray());
+						},
+						() -> {
+							s.getOwner().ifPresentOrElse(
+								m -> {
+									buf.put((byte) 1);
+									REFieldSerialization.serializeKey(buf, m);
+								},
+								() -> buf.put((byte) 0)
+							);
+						}
+					);
+					REFieldSerialization.serializeString(buf, s.getName());
+					REFieldSerialization.serializeString(buf, s.getDescription());
+					REFieldSerialization.serializeString(buf, s.getUrl());
+					REFieldSerialization.serializeString(buf, s.getIconUrl());
+				}
+			)
 		);
 
-		os.particle(
-			TokensInAccount.class,
-			ParticleDefinition.<TokensInAccount>builder()
-				.staticValidation(TokenDefinitionUtils::staticCheck)
-				.build()
+		os.substate(
+			new SubstateDefinition<>(
+				TokensInAccount.class,
+				Set.of(SubstateTypeId.TOKENS.id(), SubstateTypeId.TOKENS_LOCKED.id()),
```

Comment:
This is very personal - I didn't like the idea of reusing the same class for different types of substate, even though they share similar data structure; it introduces complexity with little benefits.

Reply:
Yes, totally agree. In this case though, TOKENS_LOCKED is going to be deprecated so would rather keep the nastiness in one place and remove it come mainnet as its not just similar data structure but also similar state machines where these two need to be handled the same and we cant use interfaces for substates at the moment.